msprime/intervals.py

```diff
@@ -229,26 +341,111 @@ def uniform_map(cls, length, rate, num_loci=None):
         return cls([0, length], [rate, 0], num_loci=num_loci)
 
     @classmethod
-    def read_hapmap(cls, filename):
+    def read_hapmap(cls, filename, sequence_length=None):
         """
-        Parses the specified file in HapMap format.
+        Parses the specified file in HapMap format. These files must be
+        white-space-delimited, and contain a single header line (which is
+        ignored) and a set of rows that pertain to a single chromosome.
+        Each line after the header is expect to contains a name (e.g. chr1),
+        and the starting position and recombination rate for the segment from
+        that position (inclusive) to the starting position on the next line
+        (exclusive). Starting positions of each segment are given in units of
+        bases, and recombination rates in centimorgans/Megabase. A warning is
+        given if the name in the first column differs between rows. Otherwise
+        only the second (Position) and third (Rate) columns are read: any
+        further columns are ignored. If the first starting position is not equal
+        to zero, then a zero-recombination region is inserted at the start of
+        the chromosome.
 
         .. warning::
             This method is deprecated, use the module-level
-            :func:`read_hapmap` function instead.
+            :func:`read_hapmap` function instead, which reads the genetic position
+            in centiMorgans from the file, rather than using the rate column.
 
         :param str filename: The name of the file to be parsed. This may be
-            in plain text or gzipped plain text.
+            in plain text or gzipped plain text. It may also be a file object.
+        :param str sequence_length: The total length of the map. If ``None``, then
+            assume it is the final physical position in the file. Otherwise pad the
+            map with a zero rate region between the final position in the file and
+            the value specified here.
         :return: A RecombinationMap object.
         """
         warnings.warn(
             "RecombinationMap.read_hapmap() is deprecated. "
             "Use msprime.read_hapmap() instead.",
             FutureWarning,
         )
-        rate_map = read_hapmap(filename)
-        rate = np.append(rate_map.rate, 0)
-        return cls(rate_map.position, rate, map_start=rate_map.map_start)
+        rownames, position, rate = np.loadtxt(
```

Comment:
It's pretty confusing having two different methods doing the same thing and having slightly different functionality, and one being deprecated. I think we'll have to support the hapmap-rate-column version indefinitely, so we can't just lump it into the deprecated RecombinationMap interface.

I think we'll need to have to think of a way to make the ``read_hapmap`` function work in both cases.